huggingface_pipelines/pipeline.py

```diff
@@ -0,0 +1,221 @@
+from abc import ABC, abstractmethod
+from typing import List, TypedDict, Dict, Any
+import logging
+from dataclasses import dataclass, replace
+from datasets import Dataset, IterableDataset
+import os
+from contextlib import contextmanager
+import torch
+from .dataset import DatasetConfig
+
+# Set up logging
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+class PipelineOverwrites(TypedDict):
+    """
+    A TypedDict representing the possible overwrites for a PipelineConfig.
+
+    Attributes:
+        batch_size (int): The new batch size to use for processing.
+        device (str): The new device to use for processing (e.g., 'cpu', 'cuda').
+        cache_to_arrow (bool): Whether to cache results to Arrow format.
+        take (int): The new number of batches to process (-1 for all).
+        output_dir (str): The new output directory for results.
+        output_file_name (str): The new output file name for results.
+        columns (List[str]): The new list of columns to process.
+    """
+
+    batch_size: int
+    device: str
+    cache_to_arrow: bool
+    take: int
+    output_dir: str
+    output_file_name: str
+    columns: List[str]
+
+
+@dataclass
+class PipelineConfig(ABC):
+    """
+    Abstract base class for pipeline configurations.
+
+    This class defines the common configuration parameters for all pipelines.
+    Specific pipeline implementations should inherit from this class and add
+    any additional configuration parameters they need.
+
+    Attributes:
+        columns (List[str]): The columns to be transformed by the pipeline.
+        dataset_config (DatasetConfig): The configuration related to loading datasets.
+        output_column_suffix (str): The suffix to append to output column names.
+        batch_size (int): The batch size to be used for processing.
+        device (str): The device to use for inference (e.g., 'cpu', 'cuda').
```

Comment:
more explanation needed as well, is it going to be used for all torch models by default ?